src/java.base/share/classes/java/lang/VirtualThread.java

```diff
@@ -546,17 +527,20 @@ private void afterYield() {
 
         // LockSupport.park/parkNanos
         if (s == PARKING || s == TIMED_PARKING) {
-            int newState = (s == PARKING) ? PARKED : TIMED_PARKED;
-            setState(newState);
+            int newState;
+            if (s == PARKING) {
+                setState(newState = PARKED);
+            } else {
+                // schedule unpark
+                assert parkTimeout > 0;
+                timeoutTask = schedule(this::unpark, parkTimeout, NANOSECONDS);
+                setState(newState = TIMED_PARKED);
```

Comment:
Just to be clear here, if the timeout expires before we can call `setState`, the unpark is basically a no-op, and we will see that we have been unparked at line 541 and set the state correctly to UNPARKED.

Reply:
Yes, and same thing is unparked by some other thread while the target thread is parking. We have several tests that bash on this.